<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>UNCURL &#8212; UNCURL 0.2.3 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Details on unsupported methods" href="unsupported_methods.html" />
    <link rel="prev" title="Welcome to UNCURL’s documentation!" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="unsupported_methods.html" title="Details on unsupported methods"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to UNCURL’s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">UNCURL 0.2.3 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="uncurl">
<h1>UNCURL<a class="headerlink" href="#uncurl" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://travis-ci.org/yjzhang/uncurl_python"><img alt="https://travis-ci.org/yjzhang/uncurl_python.svg" src="https://travis-ci.org/yjzhang/uncurl_python.svg" /></a>
<a class="reference external image-reference" href="https://pypi.python.org/pypi/uncurl-seq"><img alt="https://img.shields.io/pypi/v/uncurl-seq.svg" src="https://img.shields.io/pypi/v/uncurl-seq.svg" /></a>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>Uncurl can be installed from PyPI: <code class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">uncurl-seq</span></code>.</p>
<p>Alternatively, uncurl can be installed from source: After cloning the repository, first run <code class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">-r</span> <span class="pre">requirements.txt</span></code> to install the required libraries. Then, run <code class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">.</span></code></p>
<p>Requirements: numpy, scipy, cython, scikit-learn</p>
<p>Tested on python 2.7, 3.4 on Linux.</p>
<p>For parallel state estimation, OpenMP is required.</p>
<p>To run tests: <code class="docutils literal"><span class="pre">python</span> <span class="pre">setup.py</span> <span class="pre">test</span></code></p>
<p>Examples: see the examples folder.</p>
<p><a class="reference external" href="https://yjzhang.github.io/uncurl_python/">Full documentation</a></p>
</div>
<div class="section" id="publications">
<h2>Publications<a class="headerlink" href="#publications" title="Permalink to this headline">¶</a></h2>
<p>Accepted for ISMB 2018 (To appear in Bioinformatics).</p>
<p><a class="reference external" href="https://www.biorxiv.org/content/early/2018/03/01/142398">Preprint</a></p>
</div>
</div>
<div class="section" id="features">
<h1>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<div class="section" id="state-estimation">
<h2>State Estimation<a class="headerlink" href="#state-estimation" title="Permalink to this headline">¶</a></h2>
<p>The simplest way to use state estimation is to use the <code class="docutils literal"><span class="pre">run_state_estimation</span></code> function, which can be used to call any of the state estimation functions for different distributions. The possible distributions are &#8216;Poiss&#8217;, &#8216;LogNorm&#8217;, &#8216;Gaussian&#8217;, &#8216;NB&#8217; (negative binomial), or &#8216;ZIP&#8217; (zero-inflated Poisson). Generally, &#8216;Poiss&#8217; is recommended for sparse or count-valued datasets. Currently the NB and ZIP options are unsupported.</p>
<p>Before running state estimation, it is often a good idea to subset the number of genes. This can be done using the function <code class="docutils literal"><span class="pre">max_variance_genes</span></code>, which bins the genes by mean expression, and selects a top fraction of genes by variance from each bin. It also removes genes that have all zero expression counts.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy.io</span>
<span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">max_variance_genes</span><span class="p">,</span> <span class="n">run_state_estimation</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">)</span>

<span class="c1"># sparse data (matrix market format)</span>
<span class="n">data_sparse</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">mmread</span><span class="p">(</span><span class="s1">&#39;matrix.mtx&#39;</span><span class="p">)</span>

<span class="c1"># max variance genes, default parameters</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">max_variance_genes</span><span class="p">(</span><span class="n">data_sparse</span><span class="p">,</span> <span class="n">nbins</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">data_subset</span> <span class="o">=</span> <span class="n">data_sparse</span><span class="p">[</span><span class="n">genes</span><span class="p">,:]</span>

<span class="n">M</span><span class="p">,</span> <span class="n">W</span><span class="p">,</span> <span class="n">ll</span> <span class="o">=</span> <span class="n">run_state_estimation</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="s1">&#39;Poiss&#39;</span><span class="p">,</span> <span class="n">disp</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">max_iters</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">inner_max_iters</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">initialization</span><span class="o">=</span><span class="s1">&#39;tsvd&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="n">M2</span><span class="p">,</span> <span class="n">W2</span><span class="p">,</span> <span class="n">cost</span> <span class="o">=</span> <span class="n">run_state_estimation</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="s1">&#39;LogNorm&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="details">
<h3>Details<a class="headerlink" href="#details" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">run_state_estimation</span></code> is actually a wrapper around several other functions for state estimation.</p>
<p>The <code class="docutils literal"><span class="pre">poisson_estimate_state</span></code> function is used to estimate cell types using the Poisson Convex Mixture Model. It can take in dense or sparse matrices of reals or integers as input, and can be accelerated by parallelization. The input is of shape (genes, cells). It has three outputs: two matrices <code class="docutils literal"><span class="pre">M</span></code> and <code class="docutils literal"><span class="pre">W</span></code>, and <code class="docutils literal"><span class="pre">ll</span></code>, the negative log-likelihood. M is a (genes, clusters) matrix, and W is a (clusters, cells) matrix where each column sums to 1. The outputs <code class="docutils literal"><span class="pre">W</span></code> and <code class="docutils literal"><span class="pre">M*W</span></code> can be used for further visualization or dimensionality reduction, as described latter.</p>
<p>There are a number of different initialization methods and options for <code class="docutils literal"><span class="pre">poisson_estimate_state</span></code>. By default, it is initialized using truncated SVD + K-means, but it can also be initialized using <code class="docutils literal"><span class="pre">poisson_cluster</span></code> or just K-means.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">max_variance_genes</span><span class="p">,</span> <span class="n">poisson_cluster</span><span class="p">,</span> <span class="n">poisson_estimate_state</span>

<span class="c1"># poisson state estimation</span>
<span class="n">M</span><span class="p">,</span> <span class="n">W</span><span class="p">,</span> <span class="n">ll</span> <span class="o">=</span> <span class="n">poisson_estimate_state</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

<span class="c1"># labels in 0...k-1</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">W</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># optional arguments</span>
<span class="n">M</span><span class="p">,</span> <span class="n">W</span><span class="p">,</span> <span class="n">ll</span> <span class="o">=</span> <span class="n">poisson_estimate_state</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">disp</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">max_iters</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">inner_max_iters</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">initialization</span><span class="o">=</span><span class="s1">&#39;tsvd&#39;</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="c1"># initialization by providing means and weights</span>
<span class="n">assignments_p</span><span class="p">,</span> <span class="n">centers</span> <span class="o">=</span> <span class="n">poisson_cluster</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">M</span><span class="p">,</span> <span class="n">W</span><span class="p">,</span> <span class="n">ll</span> <span class="o">=</span> <span class="n">poisson_estimate_state</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">init_means</span><span class="o">=</span><span class="n">centers</span><span class="p">,</span> <span class="n">init_weights</span><span class="o">=</span><span class="n">assignments_p</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">log_norm_nmf</span></code> function is a wrapper around scikit-Learn&#8217;s NMF class that performs a log-transform and per-cell count normalization before running NMF. It returns two matrices, W and H, which correspond to the M and W returned by <code class="docutils literal"><span class="pre">poisson_estimate_state</span></code>. It can also take sparse matrix inputs.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">log_norm_nmf</span>

<span class="n">W</span><span class="p">,</span> <span class="n">H</span> <span class="o">=</span> <span class="n">log_norm_nmf</span><span class="p">(</span><span class="n">data_subset</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="distribution-selection">
<h2>Distribution Selection<a class="headerlink" href="#distribution-selection" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">DistFitDataset</span></code> function is used to determine the distribution of each gene in a dataset by calculating the fit error for the Poisson, Normal, and Log-Normal distributions. It currently only works for dense matrices. For large datasets, we recommend taking a small random subset of less than 1000 cells.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">DistFitDataset</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">)</span>

<span class="n">fit_errors</span> <span class="o">=</span> <span class="n">DistFitDataset</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

<span class="n">poiss_fit_errors</span> <span class="o">=</span> <span class="n">fit_errors</span><span class="p">[</span><span class="s1">&#39;poiss&#39;</span><span class="p">]</span>
<span class="n">norm_fit_errors</span> <span class="o">=</span> <span class="n">fit_errors</span><span class="p">[</span><span class="s1">&#39;norm&#39;</span><span class="p">]</span>
<span class="n">lognorm_fit_errors</span> <span class="o">=</span> <span class="n">fit_errors</span><span class="p">[</span><span class="s1">&#39;lognorm&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>The output, <code class="docutils literal"><span class="pre">fit_errors</span></code>, contains the fit error for each gene, for each of the three distributions when fitted to the data using maximum likelihood.</p>
</div>
<div class="section" id="qualitative-to-quantitative-framework">
<h2>Qualitative to Quantitative Framework<a class="headerlink" href="#qualitative-to-quantitative-framework" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">qualNorm</span></code> function is used to convert binary (or otherwise) data with shape (genes, types) into starting points for clustering and state estimation.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">qualNorm</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">)</span>
<span class="n">bin_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="s1">&#39;binary.txt&#39;</span><span class="p">)</span>
<span class="n">starting_centers</span> <span class="o">=</span> <span class="n">qualNorm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">bin_data</span><span class="p">)</span>
<span class="n">assignments</span><span class="p">,</span> <span class="n">centers</span> <span class="o">=</span> <span class="n">poisson_cluster</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="n">starting_centers</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="clustering">
<h2>Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">poisson_cluster</span></code> function does Poisson clustering with hard assignments. It takes an array of features by examples and the number of clusters, and returns two arrays: an array of cluster assignments and an array of cluster centers.</p>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">uncurl</span> <span class="kn">import</span> <span class="n">poisson_cluster</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="c1"># data is a 2d array of floats, with dimensions genes x cells</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="s1">&#39;counts.txt&#39;</span><span class="p">)</span>
<span class="n">assignments_p</span><span class="p">,</span> <span class="n">centers</span> <span class="o">=</span> <span class="n">poisson_cluster</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="dimensionality-reduction">
<h2>Dimensionality Reduction<a class="headerlink" href="#dimensionality-reduction" title="Permalink to this headline">¶</a></h2>
<p>We recommend using standard dimensionality reduction techniques such as t-SNE and PCA. They can be run on either W or <code class="docutils literal"><span class="pre">MW</span> <span class="pre">=</span> <span class="pre">M.dot(W)</span></code>. When running t-SNE on MW, we suggest taking the log and then doing a PCA or truncated SVD, as you would do for the original input data. This is the basis for the UNCURL + tSNE results in our paper. When using t-SNE on W, we suggest using a symmetric relative entropy metric, which is available as <code class="docutils literal"><span class="pre">uncurl.sparse_utils.symmetric_kld</span></code> (this can be passed in to scikit-learn&#8217;s t-SNE implementation). Cosine distance has also worked better than Euclidean distance on W.</p>
<p>Alternatively, we provide an MDS-based dimensionality reduction method that takes advantage of the convex mixture model. It is generally less accurate than t-SNE, but much faster. See <a class="reference external" href="https://yjzhang.github.io/uncurl_python/unsupported_methods.html#dimensionality-reduction">docs for unsupported methods</a>.</p>
</div>
<div class="section" id="lineage-estimation-pseudotime">
<h2>Lineage Estimation &amp; Pseudotime<a class="headerlink" href="#lineage-estimation-pseudotime" title="Permalink to this headline">¶</a></h2>
<p>The output MW of UNCURL can be used as input for other lineage estimation tools.</p>
<p>We also have implemented our own lineage estimation tools but have not thoroughly validated them. See <a class="reference external" href="https://yjzhang.github.io/uncurl_python/unsupported_methods.html#lineage-estimation">docs for unsupported methods</a>.</p>
</div>
<div class="section" id="miscellaneous">
<h2>Miscellaneous<a class="headerlink" href="#miscellaneous" title="Permalink to this headline">¶</a></h2>
<p>Unsupported methods included in the package: <a class="reference external" href="https://yjzhang.github.io/uncurl_python/unsupported_methods.html">https://yjzhang.github.io/uncurl_python/unsupported_methods.html</a></p>
<p>Miscellaneous uncurl parameters (non-default parameters and things we tried):</p>
</div>
<div class="section" id="included-datasets">
<h2>Included datasets<a class="headerlink" href="#included-datasets" title="Permalink to this headline">¶</a></h2>
<p>Real datasets:</p>
<p>10x_pooled_400.mat: 50 cells each from 8 cell types: CD19+ b cells, CD14+ monocytes, CD34+, CD56+ NK, CD4+/CD45RO+ memory t, CD8+/CD45RA+ naive cytotoxic, CD4+/CD45RA+/CD25- naive t, and CD4+/CD25 regulatory t. Source: <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/datasets">10x genomics</a>.</p>
<p>GSE60361_dat.mat: subset of data from <a class="reference external" href="http://linnarssonlab.org/cortex">Zelsel et al. 2015</a>.</p>
<p>SCDE_test.mat: data from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE29087">Islam et al. 2011</a>.</p>
<p>Synthetic datasets:</p>
<p>BranchedSynDat.mat: simulated lineage dataset with 3 branches</p>
<p>SynMouseESprog_1000.mat: simulated lineage dataset showing linear differentiation</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">UNCURL</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#publications">Publications</a></li>
</ul>
</li>
<li><a class="reference internal" href="#features">Features</a><ul>
<li><a class="reference internal" href="#state-estimation">State Estimation</a><ul>
<li><a class="reference internal" href="#details">Details</a></li>
</ul>
</li>
<li><a class="reference internal" href="#distribution-selection">Distribution Selection</a></li>
<li><a class="reference internal" href="#qualitative-to-quantitative-framework">Qualitative to Quantitative Framework</a></li>
<li><a class="reference internal" href="#clustering">Clustering</a></li>
<li><a class="reference internal" href="#dimensionality-reduction">Dimensionality Reduction</a></li>
<li><a class="reference internal" href="#lineage-estimation-pseudotime">Lineage Estimation &amp; Pseudotime</a></li>
<li><a class="reference internal" href="#miscellaneous">Miscellaneous</a></li>
<li><a class="reference internal" href="#included-datasets">Included datasets</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to UNCURL&#8217;s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="unsupported_methods.html"
                        title="next chapter">Details on unsupported methods</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/readme_link.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="unsupported_methods.html" title="Details on unsupported methods"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to UNCURL’s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">UNCURL 0.2.3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Sumit Mukherjee, Yue Zhang.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>